### ISSUE TITLE
module level setters don't work

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/Nim#14674]
module level setters don't work
### Example
```nim
when true:
  var baz0 = 0
  proc `baz=`*(a: int) =
    baz0 = a
  proc bazImpl(): int =
    baz0
  template baz*(): int =
    bazImpl()
  baz = 12
```

### Current Output
```
Error: 'bazImpl()' cannot be assigned to
```

### Expected Output
works

### Additional Information
As a motivating example, I was trying to make `system.programResult` (and `unittest.check`) work for `nim js -d:nodejs` (which despite appearances, doesn't actually work) as follows:

```nim
when defined(js) and defined(nodejs) and not defined(nimscript):
  template programResult*(): int =
    var ret: int
    asm """
`ret` = process.exitCode;
"""
    ret

  proc `programResult=`*(a: int) =
    asm """
process.exitCode = `a`;
  """
  `programResult=`(0)
```

however this is running into this issue.

But of course, there are other use cases, in particular when code evolves and you want to turn an already exposed variable into a getter/setter without breaking client code.

```
devel 1.3.5 49033eb5314d7f07e5ecc088dd0933010bb581d8
```

**Comment by Araq** [comment]
We don't support them as a setter must take two arguments.

**Comment by timotheecour** [comment]
yes, but it's "just" a matter of implementing this feature, or am I missing some technicality?

**Comment by Araq** [comment]
I thought about making modules have a type so that it becomes easier to do certain things with them like passing them to a template. Maybe it's helpful here too so that setters can keep 2 arguments.

**Comment by timotheecour** [comment]
> I thought about making modules have a type so that it becomes easier to do certain things with them like passing them to a template

that's the kind of thing enabled by https://github.com/nim-lang/Nim/pull/11992 (since a module is a symbol, you can pass it to other routines; doesn't require a type for that), and there are tests for that in that PR.

But this aside, how would it help for this issue?
I think all we need here is for sigmatch to consider `foo=` with 1 argument, this seems sound

**Comment by Araq** [comment]
> But this aside, how would it help for this issue?

As I said, then a setter still has 2 parameters, something I care about.

**Comment by timotheecour** [comment]
I understand, but this won't help with the situation described in top post, where a module level variable is turned into a getter+setter without breaking change for client code, as was the case with `system.programResult` (https://github.com/nim-lang/Nim/pull/14676):
```nim
# foo.nim
var bar = 1 # turned into a getter+setter

# client.nim
import foo
foo.bar = 2 # fully qualified can be turned into getter+setter
bar = 2 # direct access cannot
```

there are other use cases for transparently turning variable into getter+setter, such as debugging write accesses, hooks, changing implementation without changing API etc

**Comment by timotheecour** [comment]
I found an interesting solution using TRW (term rewriting macros, https://nim-lang.github.io/Nim/manual_experimental.html#term-rewriting-macros):
```nim
when true:
  var fooImpl: string
  template foo*(): string = fooImpl
  var count = 0
  template ex*{foo = val}(val: string) =
    count.inc
    fooImpl.setLen 0
    fooImpl.add val
    # fooImpl = val # this would "work" but give a truncated infinite recursion, leading to BUG1
  foo = "v1"
  doAssert foo == "v1"
  doAssert count == 1
  # echo count # BUG1: 46 instead of 1
```

however this clearly isn't ideal:
* with the simpler `fooImpl = val`, the TRW would give a truncated infinite recursion leading to BUG1; it'd end up with `foo == "v1"` but `count == 46`.
* so I had to use a hacky workaround, via `fooImpl.setLen 0; fooImpl.add val`
* this may not generalize simply to other situations

The bug here is that TRW gets evaluated in loops (until a max depth is reached, and then stops without error with its latest results) instead of detecting when a loop is found and avoiding recursing.
